### DISCUSSION TITLE
Firewall controls

### DISCUSSION THREAD

**Discussion by Msabljic** [bpg/terraform-provider-proxmox#234] [Ideas]
Would it be possible to add firewall creation or controls as either a resource or within the VM configuration?

**Comment by bpg-dev** [comment]
Hey @Msabljic 👋🏻! There is already a ticket for this: https://github.com/bpg/terraform-provider-proxmox/issues/113, which includes some implementation details.
I'm not using PVE firewall in my homelab setup at the moment, hence this is a low priority item. But PRs are always welcome! :)

**Comment by Msabljic** [comment]
Hey Pavel, funny enough I can help with that if you can create the resource
I could do testing against a working environment and give feed back. Let me
know if that helps!

**Reply by bpg-dev** [reply]
Thanks for the offer, but I don't think it would be easy to setup. I don't have a PVE instance in a cloud, and can't open external access to my homelab, sorry.

**Comment by Msabljic** [comment]
Sorry maybe I wasn’t clear, I currently have a running environment of my
own. I have been tinkering with your resources and created a few modules
with identity management and container creation. I would be interested in
helping you develop the fw resource and test the resource against my own
environment.

Let me know what you think!

**Comment by bpg-dev** [comment]
Ah, yes, sounds good! I'll try to come up with something over the weekend, and let you know when there will be something to test.
Thanks!

**Comment by Msabljic** [comment]
Cool, I’ll keep working on trying to understand the go side of it, I don’t
have a huge exposure to go lang but it seems relatively simple.

**Comment by bpg-dev** [comment]
Here is some prelim code: https://github.com/bpg/terraform-provider-proxmox/pull/246
I have security group and security group rules, along with aliases and ip sets. No in-place rule updates yet, and no binding to node / vm / lxc. 
This feature is quite big tho. I may need to take a step back to figure out how it may fit better in the terraform configs...

**Comment by bpg-dev** [comment]
Hey @Msabljic, could you give [v0.17.0-rc1](https://github.com/bpg/terraform-provider-proxmox/releases/tag/v0.17.0-rc1) a try?

**Reply by Msabljic** [reply]
Some issues I have found.

Major 
 Error: Invalid resource type
│
│   on main.tf line 16, in resource "proxmox_virtual_environment_cluster_firewall_alias" "local_network":      
│   16: resource "proxmox_virtual_environment_cluster_firewall_alias" "local_network" {
│
│ The provider bpg/proxmox does not support resource type
│ "proxmox_virtual_environment_cluster_firewall_alias".

Seems that the firewall alias module doesnt work for 0.17.0-rc1

Minor
Firewall-SG module

- ports should be numbers, however when passing a macro, port becomes blanked out meaning if any number is passed deploy fails
- passing number on protocol works but when entering gui it wants the value changed to name value. 

Firewall-options

- node_name required but not given in example

Firewall rules

- doesnt allow direct insertion of security groups, this would be an awesome thing to have so you can generate the sg's on the cluster level and pass them down to the individual resource by just giving sg-name, if it is enabled, interface and comment

I still have testing to do on other components but this is what I have found so far.

**Reply by bpg-dev** [reply]
Thanks for the feedback!

I think for alias it should be "proxmox_virtual_environment_firewall_alias" as the resource name. I must've left the old name there by accident.

I'll take a look at other comments later this week.

**Reply by bpg-dev** [reply]
> - ports should be numbers, however when passing a macro, port becomes blanked out meaning if any number is passed deploy fails

API supports range (`8080:8088`) or list (`80,443`) for ports, can't think of any alternative for handling that rather than strings.


> - passing number on protocol works but when entering gui it wants the value changed to name value.

copied that from the API docs, I think it will be better to stick with names then.

> - node_name required but not given in example

updated docs

> - doesnt allow direct insertion of security groups, this would be an awesome thing to have so you can generate the sg's on the cluster level and pass them down to the individual resource by just giving sg-name, if it is enabled, interface and comment

really good point, I totally missed this use case!


Thanks for testing!

**Reply by bpg-dev** [reply]
`0.17.0-rc2` is on the way

**Comment by Msabljic** [comment]
Yup, working on it now.

**Comment by bpg-dev** [comment]
The feature has been released in [0.17.0](https://github.com/bpg/terraform-provider-proxmox/releases/tag/v0.17.0)
